### PR TITLE
gh-121423: Improve import time of `socket` by writing `socket.errorTab` as a constant and lazy import modules

### PR DESCRIPTION
**Pull request by Wulian233**
hyperfine:
```
$ hyperfine -w 16 './python -c "import socket"'
  Time (mean ± σ):      14.6 ms ±   0.7 ms    [User: 11.9 ms, System: 2.6 ms]
  Range (min … max):    13.4 ms …  16.5 ms    192 runs
 
$ git switch pr/121424 
$ hyperfine -w 16 './python -c "import socket"'
  Time (mean ± σ):      13.9 ms ±   0.6 ms    [User: 11.4 ms, System: 2.4 ms]
  Range (min … max):    12.6 ms …  15.5 ms    197 runs
```
≈ 30% faster

* Issue: gh-121423

* Issue: gh-118761

### CODE REVIEW

**Comment by Wulian233** [comment]
Tests / Hypothesis tests on
Ubuntu failure is not my problem

File: Lib/socket.py [hunk 1]
```diff
@@ -109,103 +109,104 @@ def _intenum_converter(value, enum_klass):
...
-    errorTab[11023] = "Incorrect QoS flow count."
-    errorTab[11024] = "Unrecognized QoS object."
-    errorTab[11025] = "Invalid QoS policy object."
-    errorTab[11026] = "Invalid QoS flow descriptor."
-    errorTab[11027] = "Invalid QoS provider-specific flowspec."
-    errorTab[11028] = "Invalid QoS provider-specific filterspec."
-    errorTab[11029] = "Invalid QoS shape discard mode object."
-    errorTab[11030] = "Invalid QoS shaping rate object."
-    errorTab[11031] = "Reserved policy QoS element type."
+if sys.platform == "win32":
```
Comment (sobolevn) [inline Lib/socket.py]: This change is unrelated, I would prefer to revert it.

**Review by sobolevn — COMMENTED** [review]
I don't think that your benchmark is correct. Because you put `import socket` into your setup part. So, the import time itself is not counted.

File: Lib/socket.py (hunk 1 above)
Comment (Wulian233) [reply Lib/socket.py]: > This change is unrelated, I would prefer to revert it.
> 
> 

ok😉

**Comment by Wulian233** [comment]
```
>python -m timeit -n 1000 "import socket" "print(socket.__all__)"
1000 loops, best of 5: 424 usec per loop

>python -m timeit -n 1000 "import socket" "print(socket.__all__)"
1000 loops, best of 5: 371 usec per loop
```
1.14x faster

File: Misc/NEWS.d/next/Library/2024-07-06-12-37-10.gh-issue-121423.vnxrl4.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Speed up socket.errorTab and lazy import selectors
```
Comment (eendebakpt) [inline Misc/NEWS.d/next/Library/2024-07-06-12-37-10.gh-issue-121423.vnxrl4.rst]:
```suggestion
Improve import time of `mod`:`socket` by writing `socket.errorTab` as a constant and lazy import of `mod`:`selectors`.
```

Comment (picnixz) [reply Misc/NEWS.d/next/Library/2024-07-06-12-37-10.gh-issue-121423.vnxrl4.rst]: @eendebakpt It's <code>:mod:\`socket\`</code>, not <code>\`mode\`:\`socket\`</code> and I would suggest having <code>:data:\`socket.errorTab\`</code>, or <code>:data:\`!socket.errorTab\`</code> depending on whether the data is documented or not.

File: Misc/NEWS.d/next/Library/2024-07-06-12-37-10.gh-issue-121423.vnxrl4.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Improve import time of `mod`:`socket` by writing `socket.errorTab` as a 
+constant and lazy import of `mod`:`selectors`
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-07-06-12-37-10.gh-issue-121423.vnxrl4.rst]:
```suggestion
Improve import time of :mod:`socket` by lazy importing :mod:`selectors` and
writing :data:`!socket.errorTab` as a constant.
```

Comment (picnixz) [reply Misc/NEWS.d/next/Library/2024-07-06-12-37-10.gh-issue-121423.vnxrl4.rst]: Rewording. I know it wouldn't have made sense in the first place but I thought that the lazy import actually referred to `socket.errorTab` so with this reformulation (and Sphinx correction) I think it could be clearer.

**Comment by Wulian233** [comment]
@gpshead do you have some time to give this a review? Thanks!

**Comment by barry-scott** [comment]
The benchmark only imports socket once. the other 999 imports are no-op as socket in in sys.modules.

**Comment by Wulian233** [comment]
I carried out two parameter tests, one is with -s, only import once, is in the pr description, 1.15x. One is to do without -s and calculate the time taken to import each time, in the comments 1.14x

**Comment by effigies** [comment]
You can use [hyperfine](https://github.com/sharkdp/hyperfine) to measure the entire Python process, in which case using a `python -c pass` test is useful as a baseline:

```console
❯ hyperfine -w 16 -u microsecond 'python -c pass'
Benchmark 1: python -c pass
  Time (mean ± σ):     10921.2 µs ± 844.5 µs    [User: 8745.1 µs, System: 2091.4 µs]
  Range (min … max):   9720.3 µs … 15591.1 µs    242 runs
❯ hyperfine -w 16 -u microsecond 'python -c "import socket; socket.__all__"'
Benchmark 1: python -c "import socket; socket.__all__"
  Time (mean ± σ):     14554.3 µs ± 1154.4 µs    [User: 12061.3 µs, System: 2368.5 µs]
  Range (min … max):   12826.1 µs … 20450.0 µs    206 runs
```

**Comment by hugovk** [comment]
Using a PGO+LTO build on macOS, so this will just be measuring the `import selectors` change, and saves about 1.1 ms:

```sh
❯ hyperfine --warmup 16 \
--prepare "git checkout socket"                                   './python.exe -c "import socket; socket.__all__"' \
--prepare "git checkout 6239d41527d5977aa5d44e4b894d719bc045860e" './python.exe -c "import socket;  socket.__all__"'
Benchmark 1: ./python.exe -c "import socket; socket.__all__"
  Time (mean ± σ):      16.2 ms ±   0.8 ms    [User: 13.1 ms, System: 2.5 ms]
  Range (min … max):    15.5 ms …  21.7 ms    91 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs.

Benchmark 2: ./python.exe -c "import socket;  socket.__all__"
  Time (mean ± σ):      17.3 ms ±   1.0 ms    [User: 13.9 ms, System: 2.7 ms]
  Range (min … max):    16.6 ms …  25.1 ms    88 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs.

Summary
  ./python.exe -c "import socket; socket.__all__" ran
    1.06 ± 0.08 times faster than ./python.exe -c "import socket;  socket.__all__"
```

File: Lib/socket.py [hunk 4]
```diff
@@ -348,6 +349,8 @@ def makefile(self, mode="r", buffering=None, *,
     if hasattr(os, 'sendfile'):
 
         def _sendfile_use_sendfile(self, file, offset=0, count=None):
+            import selectors
```
Comment (hugovk) [inline Lib/socket.py:355]: Let's leave a comment to explain why, so it's not "tidied up" later:
```suggestion
            # Lazy import to improve module import time
            import selectors
```

**Review by hugovk — COMMENTED** [review]
By the way, we have a general purpose issue for import time improvements -- https://github.com/python/cpython/issues/118761 -- shall we use that as an umbrella issue for this PR as well?

---

Using [tuna](https://github.com/nschloe/tuna) to visualise import times (again, on macOS):
```sh
./python.exe -X importtime -c "import socket" 2> import.log && tuna import.log
```

Before: 3ms

image

Current PR: 2ms

image

Also: 1ms

Moving the couple of `import array`s into two functions that call it:

image

**Review by gpshead — APPROVED** [review]

File: Lib/socket.py [hunk 5]
```diff
@@ -348,6 +349,9 @@ def makefile(self, mode="r", buffering=None, *,
     if hasattr(os, 'sendfile'):
 
         def _sendfile_use_sendfile(self, file, offset=0, count=None):
+            # Lazy import to improve module import time
+            import selectors
```
Comment (vstinner) [inline Lib/socket.py:355]: I suggest to use a global variable to avoid the import at each call:

```python
global selectors
if selectors is None:
    import selectors
```

You can define the selectors variable to None at the top of the file with a comment:

```python
# module imported lazily
selectors = None
```

**Comment by vstinner** [comment]
> Also: 1ms Moving the couple of import arrays into two functions that call it:

Can you also make the array module import lazy in this PR?

**Comment by vstinner** [comment]
I measured that the change saves 0.7 ms on `import socket`: 1.72 ms => 1.24 ms.

timeit:

```
$ ./python -m timeit -s 'import sys; state=dict(sys.modules)' 'import socket; del socket; sys.modules.clear(); sys.modules.update(state)' 
200 loops, best of 5: 1.72 msec per loop

$ git switch pr/121424 
$ ./python -m timeit -s 'import sys; state=dict(sys.modules)' 'import socket; del socket; sys.modules.clear(); sys.modules.update(state)' 
200 loops, best of 5: 1.24 msec per loop
```

hyperfine:

```
$ hyperfine -w 16 './python -c "import socket"'
  Time (mean ± σ):      14.6 ms ±   0.7 ms    [User: 11.9 ms, System: 2.6 ms]
  Range (min … max):    13.4 ms …  16.5 ms    192 runs
 
$ git switch pr/121424 
$ hyperfine -w 16 './python -c "import socket"'
  Time (mean ± σ):      13.9 ms ±   0.6 ms    [User: 11.4 ms, System: 2.4 ms]
  Range (min … max):    12.6 ms …  15.5 ms    197 runs
```

**Comment by Wulian233** [comment]
Thank you! I just replaced the incorrect benchmark results in the PR with the correct ones

File: Lib/socket.py (hunk 5 above)
Comment (eendebakpt) [reply Lib/socket.py:355]: I checked several recent PRs making module loading lazy, but most of them do not use the pattern with a global variable (I suspect for readability reasons). 

@vstinner Are there any other reasons besides the small performance improvement for using the global variable?

**Comment by hugovk** [comment]
> > Also: 1ms Moving the couple of import arrays into two functions that call it:
> 
> Can you also make the array module import lazy in this PR?

@Wulian233 Please could you do this as well?

File: Lib/socket.py (hunk 5 above)
Comment (vstinner) [reply Lib/socket.py:355]: @serhiy-storchaka: Do you think that it's still useful in 2024 to use a global variable to avoid `import selectors` at each function call?

Comment (gpshead) [reply Lib/socket.py:355]: I'd just do the import. It's a mere dict lookup without a conditional when the import has already happened.

File: Lib/socket.py [hunk 6]
```diff
@@ -52,7 +52,7 @@
 import _socket
 from _socket import *
 
-import os, sys, io, selectors
+import os, sys, io
```
Comment (serhiy-storchaka) [inline Lib/socket.py]: While we are here, please split this line on separate imports.

File: Lib/socket.py (hunk 5 above)
Comment (serhiy-storchaka) [reply Lib/socket.py:355]: Just measure. It is more than a mere dict lookup (we also need to check that the module is not partially initialized, this adds 2 more dict lookups or like).

In this case, I think that the difference may be small even in comparison with a single os.fstat() call. The idiom proposed by @vstinner may be used when the whole function is very fast.

**Comment by Wulian233** [comment]
Of course! I just lazy imported array and haven't done any benchmarking yet. I also mentioned this optimization in 3.14.rst, do you think it should be included?

**Review by vstinner — APPROVED** [review]
LGTM.

What's New entry:

 > which results in a 30% speed up in standard pyperformance benchmarks.

This sentence can be misunderstood as "everything is 30% faster".

Which pyperformance benchmark is now faster?

**Comment by Wulian233** [comment]
* Improve import time of :mod:`socket` by lazy importing modules and 
   writing :data:`!socket.errorTab` as a constant, which results in 
   a 30% speed-up in the import time pyperformance benchmarks.

What do you think of this👀

**Comment by hugovk** [comment]
For What's New, we can follow the example of [3.13](https://docs.python.org/3.13/whatsnew/3.13.html#optimizations) which grouped a few import improvements together: 

> Several standard library modules have had their import times significantly improved. For example, the import time of the [typing](https://docs.python.org/3.13/library/typing.html#module-typing) module has been reduced by around a third by removing dependencies on [re](https://docs.python.org/3.13/library/re.html#module-re) and [contextlib](https://docs.python.org/3.13/library/contextlib.html#module-contextlib). Other modules to enjoy import-time speedups include [email.utils](https://docs.python.org/3.13/library/email.utils.html#module-email.utils), [enum](https://docs.python.org/3.13/library/enum.html#module-enum), [functools](https://docs.python.org/3.13/library/functools.html#module-functools), [importlib.metadata](https://docs.python.org/3.13/library/importlib.metadata.html#module-importlib.metadata), and [threading](https://docs.python.org/3.13/library/threading.html#module-threading). (Contributed by Alex Waygood, Shantanu Jain, Adam Turner, Daniel Hollas, and others in [gh-109653](https://github.com/python/cpython/issues/109653).)


Let's do the same with https://github.com/python/cpython/issues/118761 in 3.14. We have two under that issue so far.

I recommend we also group this PR under https://github.com/python/cpython/issues/118761 as well -> rename this PR title `gh-118761: ...`.

So we can follow something like that now, or leave it out for now and add a grouped summary later.

**Comment by serhiy-storchaka** [comment]
Omit details that are not interested to the end user.

I would also remove any mention from Ehat's New -- this is an insignificant change. I am sure that if you measure import time of different modules, you will find larger changes between versions (maybe even between bugfix releases) without anybody noticing.

**Comment by Wulian233** [comment]
> I recommend we also group this PR under https://github.com/python/cpython/issues/118761 as well -> rename this PR title `gh-118761: ...`.
> 
Okay, now this pr belongs to 118761. I revert the changes to 3.14 so we can write them together later when there are more modules optimizations :)

File: Misc/NEWS.d/next/Library/2024-07-06-12-37-10.gh-issue-121423.vnxrl4.rst
Comment (hugovk) [inline Misc/NEWS.d/next/Library/2024-07-06-12-37-10.gh-issue-121423.vnxrl4.rst:1]: Please rename to `Misc/NEWS.d/next/Library/2024-07-06-12-37-10.gh-issue-118761.vnxrl4.rst` (containing `gh-issue-118761`) and we're ready to merge :)

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

File: Misc/NEWS.d/next/Library/2024-07-06-12-37-10.gh-issue-121423.vnxrl4.rst
Comment (vstinner) [reply Misc/NEWS.d/next/Library/2024-07-06-12-37-10.gh-issue-121423.vnxrl4.rst:1]: https://github.com/python/cpython/issues/121423 is the right issue.

**Comment by vstinner** [comment]
Merged. Thanks for your nice enhancement @Wulian233.
